server/aesdsocket.c

```diff
@@ -0,0 +1,428 @@
+      if (ret == -1 && errno != EINTR) {
+        LOG(LOG_ERR, "recv returned -1"); perror("recv()");
+        free(recv_buf); recv_buf_size = 0; recv_buf = NULL;
+        return -1;
+      } else if (ret == 0) { // end of file (peer socket shutdown)
+        LOG(LOG_INFO, "Peer socket shutdown");
+        state = ACCEPTING_CONNECTIONS; 
+        break;
+      } else if (ret > 0) {
+        recv_buf = realloc(recv_buf, recv_buf_size + ret);
+        memcpy(recv_buf + recv_buf_size, recv_temp, ret);
+        recv_buf_size += ret;
+        char* packet_delimiter = strchr(recv_temp, '\n');
+        if (packet_delimiter != NULL) {  // delimeter found
+          state = WRITE_FILE_SOCKET_ECHO;
+        }
+      }
+      break;
+    } 
+
+    case WRITE_FILE_SOCKET_ECHO:
+    {
+      //LOG(LOG_INFO, "state: WRITE_FILE_SOCKET_ECHO");
+      // write to file
+      int tempfd = open(TEMPFILE, O_RDWR | O_CREAT | O_APPEND, 0644);
+      if (tempfd == -1) {
+        LOG(LOG_ERR, "open() returned -1"); perror("open()");
+        free(recv_buf); recv_buf_size = 0; recv_buf = NULL;
+        return -1;
+      }
+      if (-1 == write_wrapper(tempfd, recv_buf, recv_buf_size)) {
+        free(recv_buf); recv_buf_size = 0; recv_buf = NULL;
+        close(tempfd);
+        return -1;
+      } 
+
+      // realloc the recieve buffer size to 0
+      recv_buf = realloc(recv_buf, 0); recv_buf_size = 0;
+
+      // echo entire file contents to socket
+      lseek(tempfd, 0, SEEK_SET);
+      char chunk[128];
+      int nread = -1;
+      while (1) {
+        nread = read(tempfd, chunk, 128);
+        if (nread == -1 && errno != EINTR) {
+          //LOG(LOG_ERR, "nread() returned -1"); perror("read()");
+          close(tempfd);
+          return -1;
+        } else if (nread == 0) {
+          LOG(LOG_INFO, "EOF detected, socket send complete");
+          break;
+        }
+        //LOG(LOG_INFO, "forward to socket");
+        write_wrapper(peerfd, chunk, nread);
+      }
+      
+      close(tempfd);
+      state = READ_PACKET;
+      break;
+    }
+
+    default: 
+    {
+      LOG(LOG_ERR, "OMG how did I get here?");
+      return -1;
+    }
+    
+  } // end switch 
+
+  return 0;
+} // end finite_state_machine
+
+
+int write_wrapper(int fd, char* writestr, int len) 
+{
+  int written = 0;
+  while(len) {
+    written = write(fd, writestr, len);
+    if (written == -1 && errno != EINTR) {
+      LOG(LOG_ERR, "write() returned -1"); perror("write()");
+      return -1;
+    }
+    len -= written;
+    writestr += written;
+  }
+
+  return 0; 
+}
+
+char *get_ip_str(const struct sockaddr_in *sa, char *dst) 
+{
+  if (dst == NULL)
+    return NULL;
+
+  inet_ntop(AF_INET, &(sa->sin_addr), dst, INET_ADDRSTRLEN);
+  return dst;
+}
+
+
+void cleanup() 
+{
+  if (server_addr != NULL) {
+    freeaddrinfo(server_addr);
+  }
+
+  // check if file exists and if so, delete it
+  if (access(TEMPFILE, F_OK) == 0)  {
+    remove(TEMPFILE);
```

Comment:
@jmichael16  Is this function safe to do in signal handler?

Reply:
No, you are correct, remove() is not signal safe. I have fixed in future assignment when moving to the threading implementation.